File: arrays_hashing/02_valid_anagrams/solution.js
// SOLUTION 1
// class Solution {
//     /**
//      * @param {string} s
//      * @param {string} t
//      * @return {boolean}
//      */
//     isAnagram(s, t) {
//         if (s.length !== t.length) return false;
//         let sorted_s = s.split("").sort().join("")
//         let sorted_t = t.split("").sort().join("")
//         for (let i = 0; i < sorted_s.length; i++) {
//             if (sorted_s[i] !== sorted_t[i]) return false
//         }
//         return true;
//     }
// }

// value = new Solution()
// result = value.isAnagram("racecar", "carrace")
// console.log(result)

// SOLUTION 2
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {boolean}
     */
    isAnagram(s, t) {
        if (s.length !== t.length) return false;
        let hash_values = {}
        for (let i = 0; i < s.length; i++) {
            if (!(s[i] in hash_values)) hash_values[s[i]] = 0;
            hash_values[s[i]] += 1
            if (!(t[i] in hash_values)) hash_values[t[i]] = 0;
            hash_values[t[i]] -= 1
        }
        for (let value of Object.values(hash_values)) {
            if (value !== 0 ) return false
        }
        return true;
    }
}

value = new Solution()
result = value.isAnagram("racecar", "carrace")
console.log(result)

File: arrays_hashing/02_valid_anagrams/solution.py
# SOLUTION 1
# class Solution:
#     def isAnagram(self, s: str, t: str) -> bool:
#         if (len(s) != len(t)):
#             return False
#         sorted_s = "".join(sorted(s))
#         sorted_t = "".join(sorted(t))
#         for i in range(0, len(s)):
#             if (sorted_s[i] != sorted_t[i]):
#                 return False
#         return True

# value = Solution()
# result = value.isAnagram("racecar", "carrace")
# print(result)


# SOLUTION 2
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if (len(s) != len(t)):
            return False
        hashed_values = {}
        for i in range(0, len(s)):
            if (s[i] not in hashed_values):
                hashed_values[s[i]] = 0
            hashed_values[s[i]] += 1
            if (t[i] not in hashed_values):
                hashed_values[t[i]] = 0
            hashed_values[t[i]] -= 1
        for value in hashed_values.values():
            if (value != 0):
                return False
            
        return True

value = Solution()
result = value.isAnagram("racecar", "carracr")
print(result)